Independent Neopixel strips option

Bacon-Fixation opened the pull request:
### Requirements

### Description
- Add (M150 Sn) an extension of the M150 command to use it as a blend color that is different from the first NeoPixel and Caselight.
- Add LCD Menu options for control of only the the separated neopixel2 pin 

<!--

We must be able to understand your proposed change from this description. If we can't understand what the code will do from this description, the Pull Request may be closed at the maintainers' discretion. Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code recently, so please walk us through the concepts.

-->

### Benefits

- More control options on the NeoPixel pins

- Example: Say you're watching from a web cam you have the PrinterEvents on NeoPixel_pin after heat up they turn white and i want them to stay white through the print as normal but use M150 S1 to change the color of all of the LEDs on NeoPixel2_Seperate with varying intensities that are not set by the NeoPixel.
Cura adds labels to the Gcode like SHELL, INFILL,..... and doing a Search replace with M150 S1 and a color value
you can have this blending multi colored visual that let you kinda know at a glance at what phase of the layer the printer is at.

<!-- What does this fix or improve? -->

### Related Issues
- None that I know of
<!-- Whether this fixes a bug or fulfills a feature request, please list any related Issues here. -->

Bacon-Fixation committed:
Merge pull request #1 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

Bacon-Fixation committed:
seperate neopixel control with M151

sorry im not a programer but i did my best to get 2 seperated controls for extra  visual indiaction from the LEDS by keeping the Printer Events on the main channel .

Bacon-Fixation committed:
Update Configuration.h

Bacon-Fixation committed:
further intergration

Bacon-Fixation committed:
Merge branch 'bugfix-2.0.x' of https://github.com/Bacon-Fixation/Marlin into bugfix-2.0.x

Bacon-Fixation committed:
-Fixed TYPO for the Smart effector

Originally labeled as DUET_SMART_EFFECTOR
but this is incorrect as MarlinCore looks for SMART_EFFECTOR and not DUET_SMART_EFFECTOR

Bacon-Fixation commented on Marlin/Configuration.h:
```diff
@@ -946,8 +946,8 @@
-//#define DUET_SMART_EFFECTOR
-#if ENABLED(DUET_SMART_EFFECTOR)
+//#define SMART_EFFECTOR
+#if ENABLED(SMART_EFFECTOR)
```
this is misspelled when you compare it to the MarlinCore

Bacon-Fixation committed:
better intergration

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/19115

thinkyhead committed:
Neo 2 initial cleanup

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/19115

thinkyhead committed:
Clean up, extend M150

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/19115

thinkyhead committed:
Rename some pca9632 functions

thinkyhead committed:
Rename adaneo# and other vars

thinkyhead committed:
Re-use LEDColor

thinkyhead committed:
Misc. cleanup

thinkyhead committed:
Fix set_color binding

Bacon-Fixation committed:
Adding conflict protection with other neopixel features

Bacon-Fixation committed:
Merge branch 'bugfix-2.0.x' of https://github.com/Bacon-Fixation/Marlin into bugfix-2.0.x

Bacon-Fixation committed:
reverting change

Bacon-Fixation commented:
would like to still have a M151 command if possible because wanted to use it as a blend color that is different from the first NeoPixel 

for example watching from a web cam you have the PrinterEvents on NeoPixel_pin after heat up they turn white and i want them to stay white through the print as normal but use M151 to change the color of all of the LEDs on NeoPixel2_Seperate with varying intensities that are not set by the NeoPixel

Cura adds labels to the Gcode like SHELL, INFILL,..... and doing a Search replace with M151 and a color value 
you can have this blending multi colored visual that let you kinda know at a glance at what phase of the layer the printer

Bacon-Fixation committed:
fixed compiling Error with the LPC1769

Bacon-Fixation committed:
second index to stop M150 from bleeding into the second neopixel channel

thinkyhead commented:
`M150 S` is a reasonable substitute for `M151` and allows for any number of additional strips.

thinkyhead commented:
My aim in further cleanup will be to get rid of the duplicated `Marlin_NeoPixel2` class and simply make `Marlin_NeoPixel` non-static or otherwise extend it so that it can address more than one Neopixel strip. The use of `static` classes in Marlin is to make the code smaller and faster. (Non-static methods have a hidden `this` argument.) If we make `Marlin_Neopixel` into a normal non-static class then we can have multiple unique instances which are managed by the `LEDLights` class. If we keep `Marlin_Neopixel` static but add the ability for it to represent multiple LED strips, then the class becomes slightly more complicated. The best-of-all solutions is to keep `Marlin_Neopixel` static when there's only one LED strip and non-static when there's more than one. That will keep the code as lean as it can be for the most common case.

thinkyhead committed:
Who needs accessors?

thinkyhead committed:
Thanks but no thanks for your configs

thinkyhead committed:
Fix M150 to fit current design

thinkyhead committed:
Condition for conjoined strips

thinkyhead committed:
Patch neoindex

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/19115

Bacon-Fixation commented:
Sounds Great i see whats going .... cool TY again

Bacon-Fixation commented:
hey hey ran some real world testing with M150 Sn on my LPC1769 and here are my results 
NEOPIXEL_STARTUP_TEST bleeds into the second Channel
and NEOPIXEL2_STARTUP_TEST works fines
when we do a M150 S1 R127 the second channel works like it should turns on all leds in the second channel to red 
but if i run a M150 or M150 S1 half of the NeoPixels on the second channel with turn off and M150 as a hole will become unresponsive to new commands

Bacon-Fixation committed:
dont know how i lost this config

for some reason NEOPIXEL_SEPERATE ended up missing in config

Bacon-Fixation commented:
whoops still new to github

thinkyhead committed:
Update Configuration.h

thinkyhead commented:
Yes, this PR is going to be requiring a lot of tearing down and rebuilding. There will be more changes applied over the coming days, and things are bound to get broken in the process, until they get fixed. Thanks for keeping me posted about the current state.

SparkyDan555 commented:
There was a recently merged PR that (if im understanding your query correctly) will do what you are trying to achieve. See #18992 . It allows NEOPIXEL2_PIN to work independently of NEOPIXEL1_PIN by giving the second strip LED indexes of their own. Simply set NEOPIXEL2_INSERIES to true and use M150. Please let me know if this is not what you are trying to achieve.

Bacon-Fixation commented:
when i tested that build the printer event overrides pin2 when enabled and i cant set just some of the pixels of second pin it was all or nothing, but with M150 Sn we can have fully independent control of the pixels and the color that is not influenced by any other LED commands

Bacon-Fixation committed:
Update leds.h

trying out stuffs

SparkyDan555 commented:
Using the changes implemented in #18992, this is possible. Say for example you have 10 pixels on NEOPIXEL_PIN and 5 pixels on NEOPIXEL2_PIN and you wanted to set the 2nd pixel on NEOPIXEL_PIN to red. You would do this with M150 I1 R. To set the 2nd pixel on NEOPIXEL2_PIN to red you would send M150 I11 R. The only improvement I can see from your suggestions is the ability to set the entire NEOPIXEL2_PIN strip with one single command independently of the NEOPIXEL_PIN strip rather than issuing a separate M150 for every pixel on the second strip. Is this the intention?

Bacon-Fixation commented:
when i used #18992  there is no stopping printer events from influencing the second strip its still tied to pin1 indirectly
i want complete isolation from all other pin1 influence that is also separate from case lights

Bacon-Fixation committed:
Update neopixel.h

with NeoPixel2_separate enabled there isnt a conflict if they are different Neo type

Bacon-Fixation committed:
not needed

i dont know why i thought this was needed

if neopixel is't enabled then there wont be a neopixel2

Bacon-Fixation committed:
add an extra example to M150 for Sn

M150 S1 I1 R    ; Set SEPARATE index 1 to red

Bacon-Fixation committed:
Fixed Typo  for Neo2 color presets

thinkyhead committed:
Merge branch 'bugfix-2.0.x' into bugfix-2.0.x

Bacon-Fixation committed:
tinkering

I dont know why but in Conditionals_LCD.h

MULTIPLE_NEOPIXEL_TYPES  isn't carried over from its definition in neopixel.h

Bacon-Fixation committed:
NEOPIXEL_SEPERATE to run diferent neopixel types

moved the CONJOINED_NEOPIXEL from Conditionals_LCD.h to neopixel.h

because MULTIPLE_NEOPIXEL_TYPE wouldnt trip the trigger in conditionals

thinkyhead committed:
ffs

thinkyhead committed:
Update neopixel.h

thinkyhead committed:
Update leds.h

thinkyhead committed:
Update leds.h

thinkyhead committed:
Update neopixel.h

thinkyhead committed:
Update leds.cpp
